pkg/internal/tests/cross-tests/tfwrite.go

```diff
@@ -16,115 +16,70 @@
-	writeBlock(block.Body(), sch, config.AsValueMap())
-	_, err := f.WriteTo(out)
-	return err
+type hclSchemaSDKv2 struct {
+	sch map[string]*schema.Schema
 }
 
-func writeBlock(body *hclwrite.Body, schemas map[string]*schema.Schema, values map[string]cty.Value) {
-	internalMap := schema.InternalMap(schemas)
-	coreConfigSchema := internalMap.CoreConfigSchema()
+func NewHCLSchemaSDKv2(sch map[string]*schema.Schema) crosstestsimpl.ShimHCLSchema {
+	return hclSchemaSDKv2{sch: sch}
+}
 
-	blockKeys := make([]string, 0, len(coreConfigSchema.BlockTypes))
-	for key := range coreConfigSchema.BlockTypes {
-		blockKeys = append(blockKeys, key)
-	}
-	sort.Strings(blockKeys)
+var _ crosstestsimpl.ShimHCLSchema = hclSchemaSDKv2{}
 
-	for _, key := range blockKeys {
-		bl := coreConfigSchema.BlockTypes[key]
-		switch bl.Nesting.String() {
-		case "NestingSingle":
-			v, ok := values[key]
-			if !ok {
-				continue
-			}
-			newBlock := body.AppendNewBlock(key, nil)
-			res, ok := schemas[key].Elem.(*schema.Resource)
-			if !ok {
-				contract.Failf("unexpected schema type %s", key)
-			}
-			writeBlock(newBlock.Body(), res.Schema, v.AsValueMap())
-		case "NestingGroup":
-			contract.Failf("unexpected NestingGroup for %s with schema %s", key, schemas[key].GoString())
-		case "NestingList", "NestingSet":
-			v, ok := values[key]
-			if !ok {
-				continue
-			}
-			res, ok := schemas[key].Elem.(*schema.Resource)
-			if !ok {
-				contract.Failf("unexpected schema type %s", key)
-			}
-			for _, elem := range v.AsValueSlice() {
-				newBlock := body.AppendNewBlock(key, nil)
-				writeBlock(newBlock.Body(), res.Schema, elem.AsValueMap())
-			}
-		case "NestingMap":
-			contract.Failf("unexpected NestingMap for %s with schema %s", key, schemas[key].GoString())
-		default:
-			contract.Failf("unexpected nesting mode %v", bl.Nesting)
-		}
+func (s hclSchemaSDKv2) Attributes() map[string]crosstestsimpl.ShimHCLAttribute {
+	internalMap := schema.InternalMap(s.sch)
+	coreConfigSchema := internalMap.CoreConfigSchema()
+	attrMap := make(map[string]crosstestsimpl.ShimHCLAttribute)
+	for key, attr := range coreConfigSchema.Attributes {
+		attrMap[key] = attr
 	}
+	return attrMap
+}
 
-	// lifecycle block
-	if _, ok := values["lifecycle"]; ok {
-		newBlock := body.AppendNewBlock("lifecycle", nil)
-		lifecycleSchema := map[string]*schema.Schema{
-			"create_before_destroy": {
-				Type:     schema.TypeBool,
-				Optional: true,
+func (s hclSchemaSDKv2) Blocks() map[string]crosstestsimpl.ShimHCLBlock {
+	internalMap := schema.InternalMap(s.sch)
+	coreConfigSchema := internalMap.CoreConfigSchema()
+	blockMap := make(map[string]crosstestsimpl.ShimHCLBlock)
+	for key := range coreConfigSchema.BlockTypes {
+		res := s.sch[key].Elem.(*schema.Resource)
+		nesting := coreConfigSchema.BlockTypes[key].Nesting.String()
+		blockMap[key] = hclBlockSDKv2{
+			hclSchemaSDKv2: hclSchemaSDKv2{
+				sch: res.Schema,
 			},
+			nesting: sdkV2NestingToShim(nesting),
 		}
```

Comment:
We shouldn't need to re-index into the map whose keys we are already iterating over.
```suggestion
	for key, block := range coreConfigSchema.BlockTypes {
		res := s.sch[key].Elem.(*schema.Resource)
		blockMap[key] = hclBlockSDKv2{
			hclSchemaSDKv2: hclSchemaSDKv2{
				sch: res.Schema,
			},
			nesting: sdkV2NestingToShim(block.Nesting),
		}
```

Separate question: Why can't we use the schema from `block` instead of `s.sch[key].Elem.(*schema.Resource)`?

Reply:
Block doesn't contain the actual schema.Schema - it **should** contain the information we need though but we'd have to mess with the types and some of them are private. I think this is easier.